### ISSUE TITLE
%()s

### ISSUE THREAD

**Issue by larssn** [fail2ban/fail2ban#1509]
Hi, simple question.
How does %()s work? I can't find anywhere its explained.

Currently I'm trying to reference some variables in my jail.local conf, in my cloudflare action.
But `cftoken = %(cftoken)s` doesn't work.

I'm pretty much guessing when it comes to how %()s works.

**Comment by sebres** [comment]
Use `man jail.conf` or e.g. here - https://manned.org/jail.conf.5
How the default basic interpolation in config files works you can read in the python documentation (e.g. see https://docs.python.org/3/library/configparser.html#supported-ini-file-structure)

But our implementation is more complexer (allows includes, recursive interlaced substituation etc.).

Your case `cftoken = %(cftoken)s` is not allowed, because it will produce endless cycle (always substitute itself, again and again).

As well as following is not allowed also (for the same reason):
```
A = %(B)s 
B = %(C)s
C = %(A)s
```
Since we support an extensions called "known" with syntax `%(known/option)s`, after some fix (I think introduced somewhere 0.9.3 or 0.9.4, or possible first in 0.10), it allows to use the same last known option...

So try your luck with `cftoken = %(known/cftoken)s`

If you will set resp. deliver some init parameter from jail to filter or action, you can currently use another interpolation extention of us with another syntax `<option>` (Note: this is a run-time substitution):
```
# action my-action.conf
[Definition]
cftoken = <cftoken>
[Init]
cftoken = default value
```
```
# jail.local
[my-jail]
banaction = my-action[cftoken="customized value"]
```

I'll close, because should be answered thus. If not just write new comment here...

**Comment by larssn** [comment]
Thank you very much sebres. You're quite the expert.

I'll go ahead and use your last suggestion, as it seems to fit my use case the best! :)
